redash/query_runner/clickhouse.py

```
@@ -0,0 +1,120 @@
+
+        results = json.loads(results)
+
+        for row in results['rows']:
+            table_name = '{}.{}'.format(row['database'], row['table'])
+
+            if table_name not in schema:
+                schema[table_name] = {'name': table_name, 'columns': []}
+
+            schema[table_name]['columns'].append(row['name'])
+
+        return schema.values()
+
+    def _send_query(self, data, stream=False):
+        url = 'http://{host}:{port}'.format(host=self.configuration['host'], port=self.configuration['port'])
+        r = requests.post(url, data=data, stream=stream, params={
+            'user': self.configuration['user'], 'password':  self.configuration['password'],
+            'database': self.configuration['dbname']
+        })
+        if r.status_code != 200:
+            raise Exception(r.text)
+        return r.json()
+
+    @staticmethod
+    def _define_column_type(column):
+        c = column.lower()
+        if 'int' in c:
+            return TYPE_INTEGER
+        elif 'float' in c:
+            return TYPE_FLOAT
+        elif 'datetime' == c:
+            return TYPE_DATETIME
+        elif 'date' == c:
+            return TYPE_DATE
+        else:
+            return TYPE_STRING
+
+    def _clickhouse_query(self, query):
+        query += ' FORMAT JSON'
+        result = self._send_query(query)
+        columns = [{'name': r['name'], 'friendly_name': r['name'],
+                    'type': self._define_column_type(r['type'])} for r in result['meta']]
+        return {'columns': columns, 'rows': result['data']}
+
+    def run_query(self, query, user):
+        logger.info("Clickhouse is about to execute query: %s", query)
+        if query == "":
+            json_data = None
+            error = "Query is empty"
+            return json_data, error
+        try:
+            q = self._clickhouse_query(query)
+            data = json.dumps(q, cls=JSONEncoder)
+            error = None
+        except Exception as exc:
```

If we just return the exception as the `error` value, we can let it bubble up and handle it there.